internal/devices/drivers.go

```diff
@@ -0,0 +1,114 @@
+// +build windows
+
+package devices
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/Microsoft/hcsshim/internal/oci"
+	"github.com/Microsoft/hcsshim/internal/resources"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+	"github.com/pkg/errors"
+)
+
+// getAssignedDeviceUtilityTool gets the path to device-util on the server host
+func getAssignedDeviceUtilityTool(annotations map[string]string) (string, error) {
+	tools, ok := annotations[oci.AnnotationAssignedDeviceUtilityTool]
+	if !ok || tools == "" {
+		return "", errors.New("no driver tools were specified")
+	}
+	if _, err := os.Stat(tools); err != nil {
+		return "", errors.Wrapf(err, "failed to find device installation tools at %s", tools)
+	}
+	return tools, nil
+}
+
+// getAssignedDeviceKernelDrivers gets any device drivers specified on the spec.
+// Drivers are optional, therefore do not return an error if none are on the spec.
+//
+// Paths specified by this annotation should point to directories containing all
+// files and information needed to install given driver(s). This may include .sys,
+// .inf, .cer, and/or other files used during standard installation with pnputil.
+func getAssignedDeviceKernelDrivers(annotations map[string]string) ([]string, error) {
+	csDrivers, ok := annotations[oci.AnnotationAssignedDeviceKernelDrivers]
+	if !ok || csDrivers == "" {
+		return nil, nil
+	}
+	drivers := strings.Split(csDrivers, ",")
+	for _, driver := range drivers {
+		if _, err := os.Stat(driver); err != nil {
+			return nil, errors.Wrapf(err, "failed to find path to drivers at %s", driver)
+		}
+	}
+	return drivers, nil
+}
+
+// setupDeviceUtilTool finds the utility tool's host path, mounts it using vsmb, and
+// returns the UVM path to the tools
+func setupDeviceUtilTool(ctx context.Context, vm *uvm.UtilityVM, annotations map[string]string, closers []resources.ResourceCloser) (string, error) {
+	toolHostPath, err := getAssignedDeviceUtilityTool(annotations)
+	if err != nil {
+		return "", err
+	}
+	return addVSMBToUVM(ctx, vm, closers, toolHostPath)
+}
+
+// installWindowsDrivers finds specified kernel driver directories, mounts them using vsmb,
+// then installs them in the UVM
+func installWindowsDrivers(ctx context.Context, vm *uvm.UtilityVM, annotations map[string]string, closers []resources.ResourceCloser) error {
```

Comment:
I think this should just take a `[]string` of driver directory paths. The fact that we store them as an annotation is a detail that should live in `hcsoci`, since it's related to the specifics of how we map the OCI schema to our internal data model.

Reply:
Updated :)